### PR TITLE
Add Env variable for Enforcing provider

### PR DESCRIPTION
**Pull request by Revolyssup**
Signed-off-by: revolyssup <ashishjaitiwari15112000@gmail.com>

**Description**

This PR adds two functionalities - 
1.  We only had provider authentication up until this point. This PR adds authentication mechanisms as enums currently supporting only two- NoAuth and ProviderAuth. NoAuth is equivalent right now to bypassing the Auth middleware. And provider authentication is the usual None/Remote provider authentication. The reason for this approach is to get all the endpoints under AuthMiddleware for better security such that no endpoint is bare. Each endpoint can describe the type of authentication it needs while wrapping itself under AuthMiddleware. In future, for endpoints which are used by external components like adapters, we can have a different type of AuthenticationMechanism and that change will be additive here under as a new switch case expression.
2. This PR also adds a flag to enforce the provider authentication for a specific type of provider. (current use case: Enforce remote provider flag)

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: server/handlers/handler_instance.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ type Handler struct {
 	brokerConn         broker.Handler
 	K8sCompRegHelper   *models.ComponentsRegistrationHelper
 	MesheryCtrlsHelper *models.MesheryControllersHelper
+	EnforceProvider    string //If this is set, then only this provider is enforced on all endpoints
```
Comment (leecalcote) [inline server/handlers/handler_instance.go:24]:
```suggestion
	EnforceProvider    string // When set, all endpoints consider tokens / identities / capabilities valid from the single, designated provider.
```

File: server/router/server.go [hunk 2]
```diff
@@ -48,170 +48,170 @@ func NewRouter(ctx context.Context, h models.HandlerInterface, port int, g http.
...
+	gMux.Handle("/api/patterns/delete", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.DeleteMultiMesheryPatternsHandler), models.ProviderAuth))).
 		Methods("POST")
 
-	gMux.HandleFunc("/api/oam/{type}", h.OAMRegisterHandler).Methods("GET", "POST")
-	gMux.HandleFunc("/api/oam/{type}/{name}", h.OAMComponentDetailsHandler).Methods("GET")
-	gMux.HandleFunc("/api/oam/{type}/{name}/{id}", h.OAMComponentDetailByIDHandler).Methods("GET")
-	gMux.HandleFunc("/api/meshmodel/validate", h.ValidationHandler).Methods("POST")
-	gMux.HandleFunc("/api/meshmodel/component/generate", h.ComponentGenerationHandler).Methods("POST")
-	gMux.HandleFunc("/api/meshmodel/components", h.GetAllMeshmodelComponents).Methods("GET")
+	gMux.Handle("/api/oam/{type}", h.AuthMiddleware(http.HandlerFunc(h.OAMRegisterHandler), models.NoAuth)).Methods("GET", "POST")
```
Comment (leecalcote) [inline server/router/server.go:144]: Temporarily allowing NoAuth. ✔️

File: server/router/server.go [hunk 3]
```diff
@@ -48,170 +48,170 @@ func NewRouter(ctx context.Context, h models.HandlerInterface, port int, g http.
...
-	gMux.HandleFunc("/api/components", h.GetAllComponents).Methods("GET")
-	gMux.HandleFunc("/api/components/types", h.ComponentTypesHandler).Methods("GET")
-	gMux.HandleFunc("/api/components/{type}", h.ComponentsForTypeHandler).Methods("GET")
-	gMux.HandleFunc("/api/components/{type}/versions", h.ComponentVersionsHandler).Methods("GET")
-	gMux.HandleFunc("/api/components/{type}/{name}", h.ComponentsByNameHandler).Methods("GET")
+	gMux.Handle("/api/components", h.AuthMiddleware(http.HandlerFunc(h.GetAllComponents), models.NoAuth)).Methods("GET")
+	gMux.Handle("/api/components/types", h.AuthMiddleware(http.HandlerFunc(h.ComponentTypesHandler), models.NoAuth)).Methods("GET")
+	gMux.Handle("/api/components/{type}", h.AuthMiddleware(http.HandlerFunc(h.ComponentsForTypeHandler), models.NoAuth)).Methods("GET")
+	gMux.Handle("/api/components/{type}/versions", h.AuthMiddleware(http.HandlerFunc(h.ComponentVersionsHandler), models.NoAuth)).Methods("GET")
+	gMux.Handle("/api/components/{type}/{name}", h.AuthMiddleware(http.HandlerFunc(h.ComponentsByNameHandler), models.NoAuth)).Methods("GET")
```
Comment (leecalcote) [inline server/router/server.go:155]: This collection of endpoints should be protected with ProviderAuth.

File: server/router/server.go [hunk 4]
```diff
@@ -48,170 +48,170 @@ func NewRouter(ctx context.Context, h models.HandlerInterface, port int, g http.
...
+	gMux.Handle("/api/meshmodel/validate", h.AuthMiddleware(http.HandlerFunc(h.ValidationHandler), models.NoAuth)).Methods("POST")
+	gMux.Handle("/api/meshmodel/component/generate", h.AuthMiddleware(http.HandlerFunc(h.ComponentGenerationHandler), models.NoAuth)).Methods("POST")
+	gMux.Handle("/api/meshmodel/components", h.AuthMiddleware(http.HandlerFunc(h.GetAllMeshmodelComponents), models.NoAuth)).Methods("GET")
 
-	gMux.HandleFunc("/api/components", h.GetAllComponents).Methods("GET")
-	gMux.HandleFunc("/api/components/types", h.ComponentTypesHandler).Methods("GET")
-	gMux.HandleFunc("/api/components/{type}", h.ComponentsForTypeHandler).Methods("GET")
-	gMux.HandleFunc("/api/components/{type}/versions", h.ComponentVersionsHandler).Methods("GET")
-	gMux.HandleFunc("/api/components/{type}/{name}", h.ComponentsByNameHandler).Methods("GET")
+	gMux.Handle("/api/components", h.AuthMiddleware(http.HandlerFunc(h.GetAllComponents), models.NoAuth)).Methods("GET")
```
Comment (leecalcote) [inline server/router/server.go:151]: Where is the POST for this API?

Or the dedicated components/register with POST? // @sudo-NithishKarthik

**Review by leecalcote — CHANGES_REQUESTED** [review]
Unprotected endpoints need reviewed. In terms of the Capabilities Registry, only /components/register and /oam/register should be allowed with noAuth as a current exception.

File: server/router/server.go [hunk 5]
```diff
@@ -23,21 +23,21 @@ type Router struct {
 func NewRouter(ctx context.Context, h models.HandlerInterface, port int, g http.Handler, gp http.Handler) *Router {
 	gMux := mux.NewRouter()
 
-	gMux.Handle("/api/system/graphql/query", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.KubernetesMiddleware(h.MesheryControllersMiddleware(h.GraphqlMiddleware(g))))))).Methods("GET", "POST")
-	gMux.Handle("/api/system/graphql/playground", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.KubernetesMiddleware(h.MesheryControllersMiddleware(h.GraphqlMiddleware(gp))))))).Methods("GET", "POST")
+	gMux.Handle("/api/system/graphql/query", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.KubernetesMiddleware(h.MesheryControllersMiddleware(h.GraphqlMiddleware(g)))), models.ProviderAuth))).Methods("GET", "POST")
```
Comment (leecalcote) [inline server/router/server.go:26]: Are both queries and subscriptions handled under this endpoint?

**Review by leecalcote — CHANGES_REQUESTED** [review]
These docs need updated:

**API References**
- https://docs.meshery.io/reference/graphql-apis
- https://docs.meshery.io/reference/rest-apis

**Extensibility Guide**
- https://docs.meshery.io/extensibility/api
- https://docs.meshery.io/extensibility/providers

File: server/router/server.go (hunk 3 above)
Comment (Revolyssup) [reply server/router/server.go:155]: Adapter will be registering MeshModel components on this endpoint, that is why they are left to NoAuth **for now**

File: server/router/server.go (hunk 4 above)
Comment (Revolyssup) [reply server/router/server.go:151]: Not present as of now, these are being moved to /meshmodel/components where POST will be present.

File: server/router/server.go (hunk 3 above)
Comment (Revolyssup) [reply server/router/server.go:155]: Any ways, these will be put under /meshmode/components

File: server/router/server.go (hunk 5 above)
Comment (Revolyssup) [reply server/router/server.go:26]: Yes

File: server/router/server.go (hunk 3 above)
Comment (leecalcote) [reply server/router/server.go:155]: @Revolyssup, only the POST endpoint needs NoAuth, though, right?

**Comment by leecalcote** [comment]
Only two comments outstanding on this review.

**Comment by leecalcote** [comment]
> These docs need updated:
> 
> **API References**
> 
> * https://docs.meshery.io/reference/graphql-apis
> * https://docs.meshery.io/reference/rest-apis
> 
> **Extensibility Guide**
> 
> * https://docs.meshery.io/extensibility/api
> * https://docs.meshery.io/extensibility/providers

These need updated with the new "enforce" feature, including NoAuth and ProviderAuth

**Comment by l5io** [comment]
🚀 Preview for commit c1138630fb767726bf81555b65e2e2b0760639e1 at: https://636249adde83507d017d4905--meshery-docs-preview.netlify.app

**Review by leecalcote — APPROVED** [review]
Looks good.

**Pull request merged**